### DISCUSSION TITLE
PrismaClientUnknownRequestError occurs when bulk upsert is performed asynchronously

### DISCUSSION THREAD

**Discussion by root-T2** [prisma/prisma#11952] [Q&A] [answered]
I am creating a web api consisting of prisma and express. When batch registering data to db via api, using upsert raises PrismaClientUnknownRequestError.

The problem recurred even without express, so I think it is a prisma or database problem. I'm not familiar with DB so I couldn't find the cause.

When using CREATE instead of UPSERT, data registration was completed without any error. I'm using upsert because I want to update the value if possible when the key already exists in the db.

I would like to know the cause of the error and the best practices for batch inserts. Do I need to use the transaction api?

The reproduction code is shown below. I used the Quickstart project, modified script.ts.
[Quickstart: Getting started with TypeScript & SQLite | Prisma Docs](https://www.prisma.io/docs/getting-started/quickstart)

The express api receives a json consisting of `{primary_key1: data, primary_key2: data, ...}`, and the main function mimics that behavior.
Also, I am using `Object.keys(Users).forEach(async ~)`, is there a better way to write it? I have little experience writing javascript, so I don't know if this is a common way.

```typescript:scripts.ts
import { PrismaClient } from "@prisma/client";

const prisma = new PrismaClient();

function randomString() {
  return Math.random().toString(32).substring(2);
}

type User = {
  email: string;
  name: string;
};

let Users: { [key: string]: User } = {};
for (let i = 0; i < 5; i++) {
  const email = randomString();
  const name = randomString();
  Users[email] = { email: email, name: name };
}
console.log(Users);

// A `main` function so that you can use async/await
async function main() {
  Object.keys(Users).forEach(async (email) => {
    const item = Users[email];
    const x = await prisma.user.upsert({
      where: { email: item.email },
      update: {},
      create: item,
    });
  });
}

main()
  .catch((e) => {
    throw e;
  })
  .finally(async () => {
    await prisma.$disconnect();
  });

```

error

```
yarn run v1.22.17
$ ts-node ./script.ts
{
  '55ssuj6pb48': { email: '55ssuj6pb48', name: 'epkotgqb0v' },
  '9kdjc90mm1o': { email: '9kdjc90mm1o', name: 'neb4m0e9ri8' },
  tohakcavb3o: { email: 'tohakcavb3o', name: '55bmhue78p8' },
  blilqd11v28: { email: 'blilqd11v28', name: 'gfmgnrtu118' },
  c3fguv5edm: { email: 'c3fguv5edm', name: 'l6tu558j5s' }
}
/prisma_dev/starter/node_modules/@prisma/client/runtime/index.js:38692
          throw new PrismaClientUnknownRequestError(message, this.prisma._clientVersion);
                ^
PrismaClientUnknownRequestError:
Invalid `prisma.user.upsert()` invocation:


  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: ConnectionError(Timed out during query execution.) })
    at cb (/prisma_dev/starter/node_modules/@prisma/client/runtime/index.js:38692:17) {
  clientVersion: '3.7.0'
}
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by nurul3101** [comment, marked as answer]
Hey @root-T2 👋🏻 ,

Here is the working code which I have tested:

```javascript
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();

function randomString() {
  return Math.random().toString(32).substring(2);
}

type User = {
  email: string;
  name: string;
};

// A `main` function so that you can use async/await
async function main() {
  let Users: { [key: string]: User } = {};
  for (let i = 0; i < 5; i++) {
    const email = randomString();
    const name = randomString();
    Users[email] = { email: email, name: name };
  }
  console.log(Users);

  for (let i = 0; i < Object.keys(Users).length; i++) {
    const user = Users[Object.keys(Users)[i]];
    console.log('user', user);
    const response = await prisma.user.upsert({
      create: user,
      update: user,
      where: {
        email: user.email,
      },
    });

    console.log('response', response);
  }
}

main()
  .catch((e) => {
    console.error(e);
    throw e;
  })
  .finally(async () => {
    await prisma.$disconnect();
  });

```

The issue was in the forEach loop, If an asynchronous operation needs to be done then ideally `for..of` or normal `for` loop should be used. forEach loop expects a synchronous function in callback. You can read more about it [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach)
The callback function was not waiting for all the upserts to be completed and was returning beforehand and getting disconnected which caused the issue.

Please let me know if you face any other issues.

**Reply by root-T2** [reply]
Thanks, I didn't know that synchronous functions are expected in for each.
By the way, do you know why I didn't get an error when I used create instead of upsert? Is it just a coincidence that no error occurred?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
